src/scipp/compat/pandas_compat.py

```diff
@@ -59,8 +114,81 @@ def from_pandas(pd_obj: Union[pd.DataFrame, pd.Series]) -> VariableLike:
     import pandas as pd
 
     if isinstance(pd_obj, pd.DataFrame):
-        return from_pandas_dataframe(pd_obj)
+        return from_pandas_dataframe(
+            pd_obj,
+            data_columns=data_columns,
+            include_index=include_index,
+            head_parser=head_parser,
+        )
     elif isinstance(pd_obj, pd.Series):
-        return from_pandas_series(pd_obj)
+        return from_pandas_series(
+            pd_obj, include_index=include_index, head_parser=head_parser
+        )
     else:
         raise ValueError(f"from_pandas: cannot convert type '{type(pd_obj)}'")
+
+
+HeadParser = Callable[[str], Tuple[str, Optional[Unit]]]
+HeadParserArg = Optional[Union[Literal["bracket"], HeadParser]]
+
+
+def parse_bracket_head(head: str) -> Tuple[str, Optional[Unit]]:
+    """Parses strings of the form ``name [unit]``.
+
+    ``name`` may be any string that does not contain the character ``[``.
+    And ``unit`` must be a valid unit string to be parsed by ``sc.Unit(unit)``.
+    Whitespace between the name and unit is removed.
+
+    Both name and unit, including brackets, are optional.
+    If the unit is missing but empty brackets are present,
+    ``sc.units.default_unit`` is returned.
+    If the brackets are absent as well, the returned unit is ``None``.
+    This ensures that columns without unit information are not accidentally assigned
+    ``dimensionless`` which can silence downstream errors.
+
+    If the name is missing, an empty string is returned.
+
+    Parameters
+    ----------
+    head:
+        The string to parse.
+
+    Returns
+    -------
+    :
+        The parsed name and unit.
+    """
+    import re
+
+    m = re.match(r"^([^[]*)(?:\[([^[]*)])?$", head)
+    if m is None:
+        raise ValueError(
+            f"Column header does not conform to pattern `name [unit]`: {head}"
+        )
```

Comment:
Depends. It would be more tolerant in case there are individual columns which don't conform. But if the entire dataframe doesn't conform, we would silently assign no units and the user has to look into this code to understand why.

Reply:
Changed it to return `unit=None`.